File: go/e2e/blob_s3_test.go
//go:build e2e

package e2e

import (
	"bytes"
	"context"
	"strings"
	"testing"
	"time"

	goystore "github.com/goy-co/goy-store/go"
)

func TestS3BlobStore_OperationsAgainstMinIO(t *testing.T) {
	store := setupMinIOBucket(t)
	ctx := context.Background()

	key := "releases/v1.0.0/app-binary.tar.gz"
	payload := []byte{0xCA, 0xFE, 0xBA, 0xBE, 0x01, 0x02, 0x03, 0x04}
	meta := &goystore.Metadata{
		ContentType: "application/gzip",
		Custom: map[string]string{
			"version": "1.0.0",
			"commit":  "abcdef",
		},
	}

	// 1. Put
	if err := store.Put(ctx, key, payload, meta); err != nil {
		t.Fatalf("Put to S3 failed: %v", err)
	}

	// 2. Get
	fetched, fetchedMeta, exists, err := store.Get(ctx, key)
	if err != nil || !exists {
		t.Fatalf("Get from S3 failed: exists=%v, err=%v", exists, err)
	}
	if !bytes.Equal(fetched, payload) {
		t.Fatalf("data mismatch from S3")
	}
	if fetchedMeta.ContentType != "application/gzip" || fetchedMeta.Custom["version"] != "1.0.0" {
		t.Fatalf("metadata mismatch: %+v", fetchedMeta)
	}

	// 3. List
	prefix := "releases"
	list, err := store.List(ctx, &prefix)
	if err != nil || len(list) != 1 || list[0] != key {
		t.Fatalf("List failed: list=%+v, err=%v", list, err)
	}

	// 4. PresignURL
	url, err := store.PresignURL(ctx, key, 5*time.Minute)
	if err != nil || !strings.Contains(url, "http://") || !strings.Contains(url, key) {
		t.Fatalf("PresignURL failed: url=%s, err=%v", url, err)
	}

	// 5. Health Check
	health, err := store.IsHealthy(ctx)
	if err != nil || health.State != goystore.HealthHealthy || health.Backend != "s3" {
		t.Fatalf("IsHealthy failed: health=%+v, err=%v", health, err)
	}

	// 6. Delete
	if err := store.Delete(ctx, key); err != nil {
		t.Fatalf("Delete failed: %v", err)
	}

	_, _, exists, err = store.Get(ctx, key)
	if err != nil || exists {
		t.Fatalf("expected key to be deleted from S3, exists=%v, err=%v", exists, err)
	}
}

File: rust/tests/blob_s3_test.rs
use super::common::*;
use goy_store::blob::{BlobStore, Metadata};
use std::collections::HashMap;
use std::time::Duration;

#[tokio::test]
async fn test_s3_blob_store_operations_against_minio() {
    let store = setup_minio_bucket().await.expect("failed to setup minio bucket");

    let key = "releases/v1.0.0/app-binary.tar.gz";
    let payload = vec![0xCA, 0xFE, 0xBA, 0xBE, 0x01, 0x02, 0x03, 0x04];
    let mut custom = HashMap::new();
    custom.insert("version".to_string(), "1.0.0".to_string());
    custom.insert("commit".to_string(), "abcdef".to_string());

    let meta = Metadata {
        content_type: Some("application/gzip".to_string()),
        custom,
    };

    // 1. Put
    store.put(key, &payload, Some(meta.clone())).await.unwrap();

    // 2. Get
    let fetched = store.get(key).await.unwrap();
    assert!(fetched.is_some(), "expected object to exist in S3");
    let (data, fetched_meta) = fetched.unwrap();
    assert_eq!(data, payload);
    assert_eq!(fetched_meta.content_type, Some("application/gzip".to_string()));
    assert_eq!(fetched_meta.custom.get("version").map(|s| s.as_str()), Some("1.0.0"));

    // 3. List
    let list = store.list(Some("releases")).await.unwrap();
    assert_eq!(list, vec!["releases/v1.0.0/app-binary.tar.gz"]);

    // 4. Presign URL
    let presigned_url = store.presign_url(key, Duration::from_secs(300)).await.unwrap();
    assert!(presigned_url.contains("http://"));
    assert!(presigned_url.contains(key));

    // 5. Health check
    let health = store.is_healthy().await.unwrap();
    assert_eq!(health.state, goy_store::health::HealthState::Healthy);
    assert_eq!(health.contract, "blob");
    assert_eq!(health.backend, "s3");

    // 6. Delete
    store.delete(key).await.unwrap();
    let after_delete = store.get(key).await.unwrap();
    assert!(after_delete.is_none());
}
